lib/charms/layer/zookeeper.py

```diff
@@ -97,10 +111,29 @@ def start(self):
         zookeeper_home = self.dist_config.path('zookeeper')
         self.stop()
         utils.run_as('zookeeper', '{}/bin/zkServer.sh'.format(zookeeper_home), 'start')
+        self.start_rest()
 
     def stop(self):
         zookeeper_home = self.dist_config.path('zookeeper')
         utils.run_as('zookeeper', '{}/bin/zkServer.sh'.format(zookeeper_home), 'stop')
+        self.stop_rest()
+
+    def start_rest(self):
+        self.stop_rest()
+        zookeeper_rest = self.dist_config.path('zookeeper') / 'src/contrib/rest'
+        zkrest_logs = self.dist_config.path('zookeeper_log_dir') / 'rest.out'
+        zkrest_buildxml = zookeeper_rest / 'build.xml'
+        self.run_bg('zookeeper', zkrest_logs, 'nohup', 'ant', 'run', '-f', zkrest_buildxml)
+
+    def stop_rest(self):
+        print("*** Stopping rest ***")
+        try:
+            # Give some time to the java process to spawn
+            # in the case of a start followed by an immediate stop
+            sleep(30)
+            utils.run_as('root', 'pkill', '-f', 'RestMain')
```

Comment:
The sleep before stopping the service was wrong. I had to make sure the service starts properly before reporting it is ready. Knowing the REST service has started we do not need to wait for anything during when stopping the service.

In the latest commit we are pgrepping with a time out. In case we exceed the time out we report a failure. If the service comes up before the time out we detect that (almost) immediately.